src/ln/channelmonitor.rs

```diff
@@ -119,27 +126,99 @@ pub struct SimpleManyChannelMonitor<Key> {
 	chain_monitor: Arc<ChainWatchInterface>,
 	broadcaster: Arc<BroadcasterInterface>,
 	pending_events: Mutex<Vec<events::Event>>,
+	pending_htlc_updated: Mutex<HashMap<PaymentHash, Vec<(Option<PaymentPreimage>, Option<HTLCSource>)>>>,
 	logger: Arc<Logger>,
 }
 
 impl<Key : Send + cmp::Eq + hash::Hash> ChainListener for SimpleManyChannelMonitor<Key> {
 	fn block_connected(&self, header: &BlockHeader, height: u32, txn_matched: &[&Transaction], _indexes_of_txn_matched: &[u32]) {
 		let block_hash = header.bitcoin_hash();
 		let mut new_events: Vec<events::Event> = Vec::with_capacity(0);
+		let mut htlc_updated_infos = Vec::new();
 		{
 			let mut monitors = self.monitors.lock().unwrap();
 			for monitor in monitors.values_mut() {
-				let (txn_outputs, spendable_outputs) = monitor.block_connected(txn_matched, height, &block_hash, &*self.broadcaster);
+				let (txn_outputs, spendable_outputs, mut htlc_updated) = monitor.block_connected(txn_matched, height, &block_hash, &*self.broadcaster);
 				if spendable_outputs.len() > 0 {
 					new_events.push(events::Event::SpendableOutputs {
 						outputs: spendable_outputs,
 					});
 				}
+
 				for (ref txid, ref outputs) in txn_outputs {
 					for (idx, output) in outputs.iter().enumerate() {
 						self.chain_monitor.install_watch_outpoint((txid.clone(), idx as u32), &output.script_pubkey);
 					}
 				}
+				htlc_updated_infos.append(&mut htlc_updated);
+			}
+		}
+		{
+			let mut monitors = self.monitors.lock().unwrap();
+			for htlc in &htlc_updated_infos {
+				if htlc.1.is_some() {
+					for monitor in monitors.values_mut() {
+						let our_short_channel_id;
+						match monitor.key_storage {
+							Storage::Local { ref short_channel_id, ..  } => {
+								our_short_channel_id = *short_channel_id.as_ref().unwrap();
+							},
+							Storage::Watchtower { .. } => {
+								unimplemented!();
+							}
+						}
+						if let Some(ref htlc_source) = htlc.0 {
+							match htlc_source {
+								&HTLCSource::PreviousHopData(ref source) => {
+									if source.short_channel_id == our_short_channel_id {
+										monitor.provide_payment_preimage(&htlc.2, &htlc.1.unwrap());
```

Comment:
Hmm, doing the update here really isnt sufficient, though getting it right seems nontrivial. We really need to also generate a monitor update call for the user to pass the updated monitor to the watchtower. In the case this upstream channel isnt already on-chain, this is also redundant work - the claim back from ChannelManager is sufficient. Maybe we should just have a call ChannelManager can make back into the ManyChannelMonitor saying "please update the monitor with id X and also update it at all watchtowers with the given payment_preimage"? Need to make provide_payment_preimage pub, then, but I think that's ok?

Reply:
"In the case this upstream channel isn't already on-chain" but it's not gonna be called here if no payment_preimage was detected? Agree need to update watchtowers here, seems easy doable if we send back monitor id in HTLCUpdate, a likely call could be made just after fetch_pending_htlc_updated in ChannelManager block_connected, will try to write something like that

Reply:
But isn't this duplicated code - if we get a preimage from an off-chain channel and have to claim it backwards to an on-chain channel ChannelManager needs to know how to find that ChannelMonitor and update it. If we are claiming on-chain to on-chain (ie right there) that's still the same from the claim-backwards PoV, so why not just keep that logic to ChannelManager (where it may be easier to update the monitor on watchtowers).